oil_lang/expr_eval.py

```diff
@@ -1017,18 +1084,42 @@ def _EvalBinary(self, node):
     raise NotImplementedError(node.op.id)
 
   def _EvalRange(self, node):
-    # type: (expr.Range) -> Any # XXX
+    # type: (expr.Range) -> value_t
+
+    UP_lower = _PyObjToValue(self._EvalExpr(node.lower))
+    if UP_lower.tag() != value_e.Int:
+      raise error.InvalidType('Expected Int', loc.Missing)
+
+    lower = cast(value.Int, UP_lower)
+
+    UP_upper = _PyObjToValue(self._EvalExpr(node.upper))
+    if UP_upper.tag() != value_e.Int:
+      raise error.InvalidType('Expected Int', loc.Missing)
+
+    upper = cast(value.Int, UP_upper)
 
-    lower = self._EvalExpr(node.lower)
-    upper = self._EvalExpr(node.upper)
-    return xrange(lower, upper)
+    return value.Range(IntParamBox(lower.i), IntParamBox(upper.i), IntParamBox(1))
 
   def _EvalSlice(self, node):
-    # type: (expr.Slice) -> Any # XXX
+    # type: (expr.Slice) -> value_t
 
-    lower = self._EvalExpr(node.lower) if node.lower else None
-    upper = self._EvalExpr(node.upper) if node.upper else None
-    return slice(lower, upper)
+    lower = None # type: Optional[IntParamBox]
+    upper = None # type: Optional[IntParamBox]
+    if node.lower:
+      UP_lower = _PyObjToValue(self._EvalExpr(node.lower))
+      if UP_lower.tag() != value_e.Int:
+        raise error.InvalidType('Expected Int', loc.Missing)
+
+      lower = IntParamBox(cast(value.Int, UP_lower).i)
+
+    if node.upper:
+      UP_upper = _PyObjToValue(self._EvalExpr(node.upper))
+      if UP_upper.tag() != value_e.Int:
+        raise error.InvalidType('Expected Int', loc.Missing)
+
+      upper = IntParamBox(cast(value.Int, UP_upper).i)
+
+    return value.Slice(lower, upper, IntParamBox(1))
```

Comment:
is the  stride in the value.Slice()  always 1  ?

Reply:
If they're being constructed with `:`-syntax, yes. But, `slice(...)` might produce one with a different step. This is currently captured in `_PyObjToValue()`, but will eventually be a "first class" thing once we overhaul the builtins